biscuit-auth/src/token/unverified.rs

```diff
@@ -294,37 +297,44 @@ impl UnverifiedBiscuit {
     }
 
     pub fn append_third_party(&self, slice: &[u8]) -> Result<Self, error::Token> {
-        let next_keypair = KeyPair::new_with_rng(&mut rand::rngs::OsRng);
+        let next_keypair =
+            KeyPair::new_with_rng(super::builder::Algorithm::Ed25519, &mut rand::rngs::OsRng);
+        self.append_third_party_with_keypair(slice, next_keypair)
+    }
 
+    pub fn append_third_party_with_keypair(
+        &self,
+        slice: &[u8],
+        next_keypair: KeyPair,
+    ) -> Result<Self, error::Token> {
         let ThirdPartyBlockContents {
             payload,
             external_signature,
         } = schema::ThirdPartyBlockContents::decode(slice).map_err(|e| {
             error::Format::DeserializationError(format!("deserialization error: {:?}", e))
         })?;
 
-        if external_signature.public_key.algorithm != schema::public_key::Algorithm::Ed25519 as i32
-        {
-            return Err(error::Token::Format(error::Format::DeserializationError(
-                format!(
-                    "deserialization error: unexpected key algorithm {}",
-                    external_signature.public_key.algorithm
-                ),
-            )));
-        }
-        let external_key =
-            PublicKey::from_bytes(&external_signature.public_key.key).map_err(|e| {
-                error::Format::BlockSignatureDeserializationError(format!(
-                    "block external public key deserialization error: {:?}",
-                    e
-                ))
+        let algorithm =
+            Algorithm::from_i32(external_signature.public_key.algorithm).ok_or_else(|| {
+                error::Format::DeserializationError(
+                    "deserialization error: invalid external key algorithm".to_string(),
+                )
             })?;
+        let external_key = PublicKey::from_bytes(
+            &external_signature.public_key.key,
+            match algorithm {
```

Comment:
from instance would be nice here

Reply:
done in https://github.com/biscuit-auth/biscuit-rust/commit/bdb8c7414107ddb16c8cd9b1bd6d4f0f0d0240f5